sage.py

```diff
@@ -215,259 +171,271 @@ def parse(unparsed_data, alert_labels=[], slim=False):
-                 bottom=bot) 
-        p.append(plot)
+            index = [x for x in range(tid)]
+            bottom = np.add(percentages[0], percentages[1])
+            for i in index[2:]:
+                bottom = np.add(bottom, percentages[i]).tolist()
+            plot = plt.bar(indices, percentages[tid], bar_width, bottom=bottom)
+        plots.append(plot)
 
         # TODO: Decide whether to put it like this or normalize over columns
-    #print(t)
-    plt.ylabel('Percentage of occurance')
+    plt.ylabel('Percentage of occurrence')
     plt.title('Frequency of alert category')
     if SURICATA_SUMMARY:
-        plt.xticks(ind, ('c0', 'c1', 'c2', 'c3', 'c4', 'c5', 'c6', 'c7', 'c8', 'c9', 'c10', 'c11', 'c12', 'c13', 'c14'))
+        plt.xticks(indices, ('c0', 'c1', 'c2', 'c3', 'c4', 'c5', 'c6', 'c7', 'c8', 'c9', 'c10', 'c11', 'c12', 'c13', 'c14'))
     else:
-        plt.xticks(ind, [x.split('.')[1] for x in vals], rotation='vertical')
+        plt.xticks(indices, [x.split('.')[1] for x in micro_attack_stages], rotation='vertical')
     plt.tick_params(axis='x', which='major', labelsize=8)
     plt.tick_params(axis='x', which='minor', labelsize=8)
-    #plt.yticks(np.arange(0, 13000, 1000))
-    plt.legend([plot[0] for plot in p], team_labels)
+    # plt.yticks(np.arange(0, 13000, 1000))
+    plt.legend([plot[0] for plot in plots], _team_labels)
     plt.tight_layout()
-    #plt.show() 
-    return plt
+    plt.savefig('data_histogram-' + experiment_name + '.png')
+    # plt.show()
+
+
+def _group_alerts_per_team(_team_alerts):
+    """Reorganise alerts for each attacker per team"""
+    team_data = dict()
+    for tid, team in enumerate(team_alerts):
+        host_alerts = dict()  # (attacker, victim) -> alerts
+
+        for alert in team:
+            # Alert format: (diff_dt, src_ip, src_port, dst_ip, dst_port, sig, cat, host, ts, mcat)
+            src_ip, dst_ip, signature, ts, mcat = alert[1], alert[3], alert[5], alert[8], alert[9]
+            dst_port = alert[4] if alert[4] is not None else 65000
+            # Simply respect the source,dst format! (Correction: source is always source and dest always dest!)
+
+            # Say 'unknown' if the port cannot be resolved
+            dst_port = 'unknown' if (dst_port not in port_services.keys() or port_services[dst_port] == 'unknown') else port_services[dst_port]['name']
+
+            if (src_ip, dst_ip) not in host_alerts.keys() and (dst_ip, src_ip) not in host_alerts.keys():
+                host_alerts[(src_ip, dst_ip)] = []
+
+            if (src_ip, dst_ip) in host_alerts.keys():
+                host_alerts[(src_ip, dst_ip)].append((dst_ip, mcat, ts, dst_port, signature)) # TODO: remove the redundant host names
+            else:
+                host_alerts[(dst_ip, src_ip)].append((src_ip, mcat, ts, dst_port, signature))
+
+        team_data[tid] = host_alerts.items()
+    return team_data
+
+
+def _get_ups_and_downs(frequencies, slopes):
```

Comment:
I have extracted some parts from `get_episodes` method into `_get_ups_and_downs` (which was also mentioned in the paper) and `_plot_episodes`